UI code reusability: pagination

carlosthe19916 opened the pull request:
Resolves: https://github.com/trustification/trustification/issues/152

The following code tries to separate "UI Component definition (HTML)" from the "Data management (data injected in the UI component)"

- **UI Component definition (HTML)**: `SimplePagination` is a UI component that should render the Patternfly Pagination component
- **Data management (data injected in the UI component)**: `use_pagination_state` is a hook that should manage the pages state logic and avoid repeating this code multiple times.

This PR is only replacing the current logic of https://github.com/trustification/trustificatispog/ui/src/components/vulnerability.rs but the same logic could be applied to any other component that uses pagination.

> The `use_pagination_state` is promoting the usage of `Page` and `Page Size` rather than `limit` and `offset` because for UI components it makes sense to define pages and pageSizes because that is the natural language of Tables, although I'm aware that for REST APIs `limit` and `offset` might make more sense. What do you think?

This is just a draft PR and I'd would love to get your feedback guys.

lulf commented on spog/ui/src/components/vulnerability.rs:
```diff
@@ -58,20 +64,20 @@ pub fn vulnerability_search(props: &VulnerabilitySearchProperties) -> Html {
                         &SearchOptions {
-                            offset: Some(offset),
-                            limit: Some(limit),
+                            offset: Some((page - 1) * per_page),
```
Just wondering if this should be (page - 1).max(0) to ensure we can't accidentally wrap around. Alternatively use checked_sub() https://docs.rs/num/latest/num/trait.CheckedSub.html.

carlosthe19916 replied:
Thanks for the review. The `use_pagination_state` should make sure invalid pages are never propated https://github.com/trustification/trustification/pull/153/files#diff-c04258e6f855e5575e6a3be0b3de08e9e718d61cd489dadb130fd48400a65ca2R41

And I also apply your suggestion moving that small logic to a separate function here https://github.com/trustification/trustification/pull/153/files#diff-3cda560b7a8c01c60599ccf022c1af4535ccc822437dbb47db755ed6051da6afR44

lulf replied:
Looks good, nice work!

lulf reviewed (commented):
Well spotted and I think it's a good improvement. Left one comment inline around the 'page' setting.

lulf commented:
Let me know when you think this is ready for a full review and we can merge it.

carlosthe19916 commented:
@lulf I think this PR does what I wanted to implement. When you have time, I would appreciate a full review of this PR please :)

lulf commented on spog/ui/src/main.rs:
```diff
@@ -7,6 +7,7 @@ mod components;
 mod console;
 mod hooks;
 mod pages;
+mod shared;
```
What do you think about instead of creating a new toplevel module here, just putting the simple_pagination inside src/components/pagination.rs ?

lulf replied:
Yup, good idea I think. Whether you want to do that in this PR or another one I leave up to you.

lulf reviewed (commented):
Looks good to me, left a comment on the location. Curious to understand why not putting it into src/components/pagination.rs

lulf reviewed (approved):
LGTM :rocket:

lulf commented:
@carlosthe19916 Can you rebase this branch with the latest changes from origin/main? The merging seems to encounter a conflict.

carlosthe19916 commented:
> @carlosthe19916 Can you rebase this branch with the latest changes from origin/main? The merging seems to encounter a conflict.

Sorry, it must've been something I did with my IDE. Now I made sure I've rebased latest changes from `main` branch into my working branch

lulf commented:
@carlosthe19916 It is still conflicting from what I can tell. I tried to rebase your branch manually to main:

```
[lulf@pteppic trustification]$ git checkout paginationState
branch 'paginationState' set up to track 'carlos/paginationState'.
Switched to a new branch 'paginationState'
[lulf@pteppic trustification]$ git rebase origin/main
Auto-merging spog/ui/src/components/advisory.rs
CONFLICT (content): Merge conflict in spog/ui/src/components/advisory.rs
Auto-merging spog/ui/src/components/vulnerability.rs
CONFLICT (content): Merge conflict in spog/ui/src/components/vulnerability.rs
Auto-merging spog/ui/src/utils/mod.rs
CONFLICT (content): Merge conflict in spog/ui/src/utils/mod.rs
error: could not apply ddcc2c6... Enhance page validation and apply pagination to vexination.rs and vulnerability.rs
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply ddcc2c6... Enhance page validation and apply pagination to vexination.rs and vulnerability.rs
```

carlosthe19916 commented:
> @carlosthe19916 It is still conflicting from what I can tell. I tried to rebase your branch manually to main:
> 
> ```
> [lulf@pteppic trustification]$ git checkout paginationState
> branch 'paginationState' set up to track 'carlos/paginationState'.
> Switched to a new branch 'paginationState'
> [lulf@pteppic trustification]$ git rebase origin/main
> Auto-merging spog/ui/src/components/advisory.rs
> CONFLICT (content): Merge conflict in spog/ui/src/components/advisory.rs
> Auto-merging spog/ui/src/components/vulnerability.rs
> CONFLICT (content): Merge conflict in spog/ui/src/components/vulnerability.rs
> Auto-merging spog/ui/src/utils/mod.rs
> CONFLICT (content): Merge conflict in spog/ui/src/utils/mod.rs
> error: could not apply ddcc2c6... Enhance page validation and apply pagination to vexination.rs and vulnerability.rs
> hint: Resolve all conflicts manually, mark them as resolved with
> hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
> hint: You can instead skip this commit: run "git rebase --skip".
> hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
> Could not apply ddcc2c6... Enhance page validation and apply pagination to vexination.rs and vulnerability.rs
> ```

My apologies for causing too much trouble with with simple task of `rebase`.
I've just rebased again and fixed conflicts. Now when I execute `git rebase origin/main` this is what I get

```
[cferiavi@dhcp-10-40-4-237 trustification]$ git checkout paginationState 
Switched to branch 'paginationState'
Your branch is up to date with 'origin/paginationState'.
[cferiavi@dhcp-10-40-4-237 trustification]$ git rebase origin/main
Current branch paginationState is up to date.
```

lulf commented:
@carlosthe19916 No worries, the GitHub behaviour around merge queues are still a bit confusing, no error shown :shrug:
